unixfs/src/dir/walk.rs

```diff
@@ -0,0 +1,777 @@
+use cid::Cid;
+
+#[derive(Debug)]
+pub struct Walker {
+    current: InnerEntry,
+    /// On the next call to `continue_walk` this will be the block, unless we have an ongoing file
+    /// walk in which case we shortcircuit to continue it.
+    next: Option<(Cid, String, usize)>,
+    pending: Vec<(Cid, String, usize)>,
+    // tried to recycle the names but that was consistently at fast and used more memory than just
+    // cloning the strings
+}
+
+fn convert_link(
+    depth: usize,
+    nth: usize,
+    link: PBLink<'_>,
+) -> Result<(Cid, String, usize), InvalidCidInLink> {
+    let hash = link.Hash.as_deref().unwrap_or_default();
+    let cid = match Cid::try_from(hash) {
+        Ok(cid) => cid,
+        Err(e) => return Err(InvalidCidInLink::from((nth, link, e))),
+    };
+    let name = match link.Name {
+        Some(Cow::Borrowed(s)) if !s.is_empty() => s.to_owned(),
+        None | Some(Cow::Borrowed(_)) => todo!("link cannot be empty"),
+        Some(Cow::Owned(s)) => s,
+    };
+    assert!(!name.contains('/'));
+    Ok((cid, name, depth))
+}
+
+fn convert_sharded_link(
+    depth: usize,
+    nth: usize,
+    link: PBLink<'_>,
+) -> Result<(Cid, String, usize), InvalidCidInLink> {
+    let hash = link.Hash.as_deref().unwrap_or_default();
+    let cid = match Cid::try_from(hash) {
+        Ok(cid) => cid,
+        Err(e) => return Err(InvalidCidInLink::from((nth, link, e))),
+    };
+    let (depth, name) = match link.Name {
+        Some(Cow::Borrowed(s)) if s.len() > 2 => (depth, s[2..].to_owned()),
+        Some(Cow::Borrowed(s)) if s.len() == 2 => (depth - 1, String::from("")),
+        None | Some(Cow::Borrowed(_)) => todo!("link cannot be empty"),
+        Some(Cow::Owned(s)) => {
+            if s.len() == 2 {
+                (depth - 1, String::from(""))
+            } else {
+                assert!(s.len() > 2);
+                (depth, s[2..].to_owned())
+            }
+        },
+    };
+    assert!(!name.contains('/'));
+    Ok((cid, name, depth))
+}
+
+impl Walker {
+    pub fn start<'a>(data: &'a [u8], cache: &mut Option<Cache>) -> Result<Walk<'a>, Error> {
+        let flat = FlatUnixFs::try_from(data)?;
+        let metadata = FileMetadata::from(&flat.data);
+
+        match flat.data.Type {
+            UnixFsType::Directory => {
+                let inner = InnerEntry::new_root_dir(metadata);
+
+                let links = flat.links
+                    .into_iter()
+                    .enumerate()
+                    // 2 == number of ancestors this link needs to have on the path, this is after
+                    // some trial and error so not entirely sure why ... ancestors always includes
+                    // the empty root in our case.
+                    .map(|(nth, link)| convert_link(2, nth, link));
+
+                Self::walk_directory(links, inner)
+            },
+            UnixFsType::HAMTShard => {
+                let inner = InnerEntry::new_root_bucket(metadata);
```

Comment:
```suggestion
                ShardedLookup::check_supported(&mut flat)?;
                let inner = InnerEntry::new_root_bucket(metadata);
```

the same needs to be added to the `continue_walk`. The "validation" could also return a value which knows how to convert the links, but at the moment there's only a single configuration for the HAMT dirs.

Reply:
Leaving this unresolved to remember this.

Reply:
Done done done.